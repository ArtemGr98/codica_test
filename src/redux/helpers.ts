export const findIndexCity = (cities: { name: string }[], name: string) => {
    return cities.length ? cities.findIndex(city => city.name === name) : -1
}

export const getLocalStorageCities = () => {
    let cities: { name: string }[] = []

    if (localStorage.getItem('cities')) {
        cities = JSON.parse(localStorage.getItem('cities') as string)
    }
    return cities
}