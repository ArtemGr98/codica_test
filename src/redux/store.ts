import { configureStore } from '@reduxjs/toolkit'
import citiesReducer from './citiesSlice'

export const store = configureStore({
    reducer: {
        cities: citiesReducer
    },
})

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;

