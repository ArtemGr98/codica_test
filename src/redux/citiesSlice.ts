import {createAsyncThunk, createSlice, PayloadAction} from '@reduxjs/toolkit'
import {cityWeatherApi, CityWeatherResponseI} from '../api/cityWeatherApi'
import {RootState} from './store'
import {AxiosError} from 'axios'
import {findIndexCity, getLocalStorageCities} from './helpers'

export interface CitiesStateI {
    loading: boolean,
    error: undefined | string,
    apiKey: string,
    units: string,
    citiesCards: CityWeatherResponseI[]
}

const initialState: CitiesStateI = {
    loading: false,
    error: undefined,
    apiKey: '1b0a340168db93a5a6d8f9db271a81ed',
    units: 'metric',
    citiesCards: []
}

export const getCityWeatherAsync = createAsyncThunk<CityWeatherResponseI,
    string, { state: RootState, rejectValue: string }>(
        'cities/getCityWeatherAsync',
        async (city, {getState, rejectWithValue}) => {
            try {
                const {units, apiKey} = getState().cities
                const response = await cityWeatherApi.getCityWeather(city, units, apiKey)

                const cities = getLocalStorageCities()
                const isNewCity = findIndexCity(cities, response.name)

                if (isNewCity === -1) {
                    cities.push({name: response.name})
                    localStorage.setItem('cities', JSON.stringify(cities))
                }
                return response
            } catch (err) {
                const error: AxiosError<string> = err as any
                return rejectWithValue(error.message)
            }
        }
    )

export const citiesSlice = createSlice({
    name: 'cities',
    initialState,
    reducers: {
        removeCity: (state, action: PayloadAction<string>) => {
            const removeCityIndex = findIndexCity(state.citiesCards, action.payload)
            state.citiesCards.splice(removeCityIndex, 1)

            const cities = getLocalStorageCities()
            const removeLSIndex = findIndexCity(cities, action.payload)
            cities.splice(removeLSIndex, 1)
            cities.length ?
                localStorage.setItem('cities', JSON.stringify(cities)) :
                localStorage.removeItem('cities')
        }
    },
    extraReducers: (builder) => {
        builder
            .addCase(getCityWeatherAsync.pending, (state) => {
                state.loading = true
                state.error = undefined
            })
            .addCase(getCityWeatherAsync.fulfilled, (state, {payload}) => {
                state.loading = false

                const isNewCity = findIndexCity(state.citiesCards, payload.name)
                isNewCity === -1 ?
                    state.citiesCards.push({
                        name: payload.name,
                        main: {
                            temp: payload.main.temp,
                            humidity: payload.main.humidity,
                            pressure: payload.main.pressure,
                            feels_like: payload.main.feels_like
                        },
                        sys: {country: payload.sys.country},
                        weather: [{
                            main: payload.weather[0].main,
                            description: payload.weather[0].description
                        }]
                    }) :
                    state.citiesCards[isNewCity] = {
                        name: payload.name,
                        main: {
                            temp: payload.main.temp,
                            humidity: payload.main.humidity,
                            pressure: payload.main.pressure,
                            feels_like: payload.main.feels_like
                        },
                        sys: {country: payload.sys.country},
                        weather: [{
                            main: payload.weather[0].main,
                            description: payload.weather[0].description
                        }]
                    }
            })
            .addCase(getCityWeatherAsync.rejected, (state, action) => {
                state.loading = false
                state.error = action.payload
            })
    },
})




export const {removeCity} = citiesSlice.actions
export default citiesSlice.reducer
