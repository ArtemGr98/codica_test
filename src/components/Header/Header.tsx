import {ChangeEvent, FC, useState} from 'react'
import {useAppDispatch} from '../../hooks/hooks'
import {getCityWeatherAsync} from '../../redux/citiesSlice'
import {AppBar, Box, Button, TextField, Toolbar} from '@mui/material'

const Header: FC = () => {
    const [searchCity, setSearchCity] = useState('')
    const dispatch = useAppDispatch()

    const handleAddCity = (e: ChangeEvent<HTMLFormElement>) => {
        e.preventDefault()
        dispatch(getCityWeatherAsync(searchCity))
        setSearchCity('')
    }

    return (
        <AppBar color="default">
            <Toolbar sx={{justifyContent: 'center'}}>
                <form onSubmit={handleAddCity}>
                    <Box sx={{display: 'flex', alignItems: 'stretch'}}>
                        <TextField type="text" label="enter city"
                            value={searchCity}
                            size="small" color="info"
                            onChange={(e) => setSearchCity(e.target.value)} />
                        <Button type="submit" color="inherit"
                            sx={{marginLeft: '10px'}}
                            disabled={!searchCity}>add city</Button>
                    </Box>
                </form>
            </Toolbar>
        </AppBar>
    )
}

export default Header