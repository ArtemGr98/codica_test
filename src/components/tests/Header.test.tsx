import {fireEvent, render, screen} from '@testing-library/react'
import Header from '../Header/Header'
import {mockedDispatch} from './helpers'
import * as actions from '../../redux/citiesSlice'

describe('Header', () => {
    it('Header render', () => {
        const view = render(<Header/>)

        expect(view).toMatchSnapshot()
    })
    it('add city', () => {
        const dispatch = jest.fn()
        mockedDispatch.mockReturnValue(dispatch)
        const mockedAddCity = jest.spyOn(actions, 'getCityWeatherAsync')

        render(<Header/>)

        const input = screen.getByRole('textbox')

        expect(input).toHaveValue('')
        fireEvent.input(input, {target: {value: 'city'}})
        expect(input).toHaveValue('city')

        fireEvent.click(screen.getByRole('button'))

        expect(dispatch).toHaveBeenCalledTimes(1)
        expect(mockedAddCity).toHaveBeenCalledWith('city')
        expect(input).toHaveValue('')

        fireEvent.click(screen.getByRole('button'))
        expect(dispatch).toHaveBeenCalledTimes(1)
    })
})