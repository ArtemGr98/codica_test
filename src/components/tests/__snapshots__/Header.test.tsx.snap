// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Header render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorDefault MuiAppBar-positionFixed mui-fixed css-1ogdjvx-MuiPaper-root-MuiAppBar-root"
      >
        <div
          class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-y44dvl-MuiToolbar-root"
        >
          <form>
            <div
              class="MuiBox-root css-19t2ho9"
            >
              <div
                class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorInfo MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined css-xewhx1-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                  for=":r0:"
                  id=":r0:-label"
                >
                  enter city
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorInfo MuiInputBase-formControl MuiInputBase-sizeSmall css-hgkgee-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
                    id=":r0:"
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-yjsfm1"
                    >
                      <span>
                        enter city
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorInherit Mui-disabled MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorInherit css-73c6kf-MuiButtonBase-root-MuiButton-root"
                disabled=""
                tabindex="-1"
                type="submit"
              >
                add city
              </button>
            </div>
          </form>
        </div>
      </header>
    </div>
  </body>,
  "container": <div>
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorDefault MuiAppBar-positionFixed mui-fixed css-1ogdjvx-MuiPaper-root-MuiAppBar-root"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-y44dvl-MuiToolbar-root"
      >
        <form>
          <div
            class="MuiBox-root css-19t2ho9"
          >
            <div
              class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorInfo MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined css-xewhx1-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for=":r0:"
                id=":r0:-label"
              >
                enter city
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorInfo MuiInputBase-formControl MuiInputBase-sizeSmall css-hgkgee-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
                  id=":r0:"
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-yjsfm1"
                  >
                    <span>
                      enter city
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorInherit Mui-disabled MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorInherit css-73c6kf-MuiButtonBase-root-MuiButton-root"
              disabled=""
              tabindex="-1"
              type="submit"
            >
              add city
            </button>
          </div>
        </form>
      </div>
    </header>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
